framework_agreement/model/framework_agreement.py

```diff
@@ -311,72 +316,56 @@ def create(self, vals):
-                 '&',
-                 ('start_date', '>=', agreement.start_date),
-                 ('start_date', '<=', agreement.end_date),
-                 '&',
-                 ('end_date', '>=', agreement.start_date),
-                 ('end_date', '<=', agreement.end_date)]
-            )
-            # we also look for the one that includes current offer
-            overlap += self.search(
-                [('start_date', '<=', agreement.start_date),
-                 ('end_date', '>=', agreement.end_date),
-                 ('id', '!=', agreement.id),
-                 ('product_id', '=', agreement.product_id.id)]
-            )
-            overlap = self.browse([x.id for x in overlap
-                                   if x.id != agreement.id])
-            # we ensure that there is only one agreement at time per product
-            # if strict agreement is set on company
-            if strict and overlap:
-                raise exceptions.Warning(
-                    _('There is already is a running agreement for '
-                      'product %s')) % agreement.product_id.name
-            # We ensure that there are not multiple agreements
-            # for same supplier at same time
-            supplier = next(
-                (x for x in overlap
-                 if x.supplier_id.id == agreement.supplier_id.id),
-                None)
-            if supplier:
-                raise exceptions.Warning(
-                    _('There can not be multiple agreement '
-                      'for supplier %s') % supplier.name
-                )
-        return True
-
-    # 'supplier_id', 'product_id',
-    # 'start_date', 'end_date' may be only watch
-    @api.multi
-    @api.constrains('supplier_id', 'product_id', 'start_date', 'end_date')
+    @api.one
+    @api.constrains('supplier_id', 'product_id', 'start_date', 'end_date',
+                    'company_id', 'incoterm_id', 'incoterm_address')
     def check_overlap(self):
-        """Constraint to check that no agreements for same product/supplier overlap.
+        """Check that there are no similar agreements at the same time.
 
-        One agreement per product limit is checked if one_agreement_per_product
-        is set to True on company
+        Depending on the one_agreement_per_product flag on the company,
+        agreements from different companies are tolerated or not.
 
         """
-        return self._check_overlap()
+        date_domain = [
+            '|',
+            '|',
+            '&',
+            # start date in between
+            ('start_date', '>=', self.start_date),
+            ('start_date', '<=', self.end_date),
+            '&',
+            # end date in between
+            ('end_date', '>=', self.start_date),
+            ('end_date', '<=', self.end_date),
+            '&',
+            # includes the current self
+            ('start_date', '<=', self.start_date),
+            ('end_date', '>=', self.end_date),
+        ]
```

Comment:
if you have a constraint that end_date > start_date, then you can remove line 341-343 (and line 330).